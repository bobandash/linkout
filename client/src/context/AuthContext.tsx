import {
  useEffect,
  useState,
  createContext,
  SetStateAction,
  Dispatch,
} from 'react';
import { FC } from 'react';
import axios from 'axios';
interface SignedInContextProviderProps {
  children: React.ReactNode;
}

interface AuthContextProps {
  isSignedIn: boolean;
  setIsSignedIn: Dispatch<SetStateAction<boolean>>;
  isAuthLoading: boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  isSignedIn: false,
  setIsSignedIn: () => {},
  isAuthLoading: true,
});

// Contains all relevant information for the dashboard
export const AuthContextProvider: FC<SignedInContextProviderProps> = ({
  children,
}) => {
  const [isSignedIn, setIsSignedIn] = useState(false);
  const [isAuthLoading, setIsAuthLoading] = useState(true);
  useEffect(() => {
    async function getSignedInStatus() {
      try {
        const response = await axios.get('/api/users/user/sign-in-status', {
          withCredentials: true,
        });
        if (response.status === 200) {
          setIsSignedIn(true);
        }
      } catch {
        setIsSignedIn(false);
      } finally {
        setIsAuthLoading(false);
      }
    }
    getSignedInStatus();
  }, []);

  return (
    <AuthContext.Provider
      value={{
        setIsSignedIn,
        isSignedIn,
        isAuthLoading,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};
