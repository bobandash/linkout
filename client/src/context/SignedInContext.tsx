import {
  useEffect,
  useState,
  createContext,
  SetStateAction,
  Dispatch,
} from 'react';
import { FC } from 'react';
import axios from 'axios';

interface SignedInContextProviderProps {
  children: React.ReactNode;
}

interface SignedInContextProps {
  isSignedIn: boolean;
  setIsSignedIn: Dispatch<SetStateAction<boolean>>;
}

export const SignedInContext = createContext<SignedInContextProps>({
  setIsSignedIn: () => {},
  isSignedIn: false,
});

export const SignedInContextProvider: FC<SignedInContextProviderProps> = ({
  children,
}) => {
  const [isSignedIn, setIsSignedIn] = useState(false);
  useEffect(() => {
    async function getSignedInStatus() {
      try {
        const response = await axios.get('/api/users/user/sign-in-status', {
          withCredentials: true,
        });
        if (response.status === 200) {
          setIsSignedIn(true);
        }
      } catch {
        setIsSignedIn(false);
      }
    }
    getSignedInStatus();
  }, []);

  return (
    <SignedInContext.Provider value={{ setIsSignedIn, isSignedIn }}>
      {children}
    </SignedInContext.Provider>
  );
};
