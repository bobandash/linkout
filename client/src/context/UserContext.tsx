import {
  useEffect,
  useState,
  createContext,
  SetStateAction,
  Dispatch,
} from 'react';
import { FC } from 'react';
import axios from 'axios-config';
import socket from '../socket';
interface SignedInContextProviderProps {
  children: React.ReactNode;
}

interface CommunitiesProp {
  name: string;
  profilePic: string;
  _id: string;
  description: string;
}

interface UserContextProps {
  isSignedIn: boolean;
  setIsSignedIn: Dispatch<SetStateAction<boolean>>;
  username: string;
  isLoading: boolean;
  communities: CommunitiesProp[] | null;
}

export const UserContext = createContext<UserContextProps>({
  setIsSignedIn: () => {},
  isSignedIn: false,
  username: '',
  isLoading: true,
  communities: null,
});

export const UserContextProvider: FC<SignedInContextProviderProps> = ({
  children,
}) => {
  const [isSignedIn, setIsSignedIn] = useState(false);
  const [username, setUsername] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const [communities, setCommunities] = useState<null | CommunitiesProp[]>(
    null,
  );

  socket.on('addServerIconSidebar', (community: CommunitiesProp) => {
    if (communities !== null) {
      setCommunities([...communities, community]);
    }
  });

  useEffect(() => {
    async function main() {
      await getUser();
      await getUsername();
      await getUserCommunities();
      setIsLoading(false);
    }

    async function getUserCommunities() {
      try {
        const response = await axios.get(
          'https://linkout.onrender.com/users/user/community',
          {
            withCredentials: true,
          },
        );
        const json = response.data;
        setCommunities(json.communities);
      } catch (err) {
        setCommunities(null);
      }
    }

    async function getUsername() {
      try {
        const response = await axios.get(
          'https://linkout.onrender.com/users/user/username',
          {
            withCredentials: true,
          },
        );
        const json = response.data;
        setUsername(json.username);
      } catch {
        setUsername('');
      }
    }

    async function getUser() {
      try {
        const response = await axios.get(
          'https://linkout.onrender.com/users/user/sign-in-status',
          {
            withCredentials: true,
          },
        );
        if (response.status === 200) {
          setIsSignedIn(true);
        }
      } catch {
        setIsSignedIn(false);
      }
    }
    main();
  }, [isSignedIn]);

  return (
    <UserContext.Provider
      value={{ setIsSignedIn, isSignedIn, username, isLoading, communities }}
    >
      {children}
    </UserContext.Provider>
  );
};
