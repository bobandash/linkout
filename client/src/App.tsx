import SignInForm from './pages/SignInForm/index';
/* import { UserContext } from './context/UserContext'; */
import { useUserContext } from './context/useUserContext';
import PageTemplate from './components/PageTemplate';
import { Outlet } from 'react-router';
import LoadingScreen from './pages/Loading/index';

function App() {
  const { isSignedIn, isLoading } = useUserContext();
  if (isLoading) {
    return <LoadingScreen />;
  }

  if (isSignedIn) {
    return (
      <PageTemplate>
        <Outlet />
      </PageTemplate>
    );
  }

  return <SignInForm />;
}

export default App;
