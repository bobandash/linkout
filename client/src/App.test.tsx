import { render, screen } from '@testing-library/react';
import App from './App';
import { UserContextProvider } from './context/UserContext';
import { MemoryRouter } from 'react-router';
import * as useUserContextModule from './context/useUserContext';

vi.mock('./context/useUserContext', () => ({
  useUserContext: vi.fn(),
}));

describe('App', () => {
  describe('Not Signed In', () => {
    beforeEach(async () => {
      const mockedUseUserContext = vi.spyOn(
        useUserContextModule,
        'useUserContext',
      );
      mockedUseUserContext.mockReturnValue({
        isSignedIn: false,
        setIsSignedIn: vi.fn(),
        username: '',
        communities: [],
        isLoading: false,
      });

      render(
        <MemoryRouter>
          <UserContextProvider>
            <App />
          </UserContextProvider>
        </MemoryRouter>,
      );
    });

    it('renders SignInForm if not signed in', () => {
      expect(
        screen.getByRole('button', { name: 'Sign In' }),
      ).toBeInTheDocument();
    });

    afterEach(() => {
      vi.clearAllMocks();
      vi.resetAllMocks();
    });
  });

  describe('Loading Screen', () => {
    beforeEach(async () => {
      const mockedUseUserContext = vi.spyOn(
        useUserContextModule,
        'useUserContext',
      );
      mockedUseUserContext.mockReturnValue({
        isSignedIn: false,
        setIsSignedIn: vi.fn(),
        username: '',
        communities: [],
        isLoading: true,
      });
    });

    it('renders Loading screen if loading state is true', () => {
      const { container } = render(
        <MemoryRouter>
          <UserContextProvider>
            <App />
          </UserContextProvider>
        </MemoryRouter>,
      );
      expect(container).toMatchSnapshot();
    });

    afterEach(() => {
      vi.clearAllMocks();
      vi.resetAllMocks();
    });
  });

  describe('Application Screen', () => {
    beforeEach(async () => {
      const mockedUseUserContext = vi.spyOn(
        useUserContextModule,
        'useUserContext',
      );
      mockedUseUserContext.mockReturnValue({
        isSignedIn: true,
        setIsSignedIn: vi.fn(),
        username: '',
        communities: [],
        isLoading: false,
      });
    });

    it('renders application is user is signed in', () => {
      const { container } = render(
        <MemoryRouter>
          <UserContextProvider>
            <App />
          </UserContextProvider>
        </MemoryRouter>,
      );
      expect(container).toMatchSnapshot();
    });

    afterEach(() => {
      vi.clearAllMocks();
      vi.resetAllMocks();
    });
  });
});
