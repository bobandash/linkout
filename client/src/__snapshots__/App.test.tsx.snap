// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Application Screen > renders application is user is signed in 1`] = `
<div>
  <div
    class="grid-cols-default_site xl:grid-cols-xl_site grid-rows-mobile_site grid overflow-hidden lg:grid-cols-lg_site 2xl:grid-cols-two_xl_site"
  >
    <aside
      class="hidden min-h-screen max-w-full flex-row overflow-hidden lg:flex"
    >
      <div
        class="flex max-h-screen flex-col items-center gap-2 overflow-scroll bg-color_1 p-4 scrollbar-none"
      >
        <div
          class="flex aspect-square w-full min-w-[50px] items-center justify-center rounded-full border-2 bg-white hover:cursor-pointer lg:min-w-[60px]"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-house "
            data-icon="house"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M575.8 255.5c0 18-15 32.1-32 32.1h-32l.7 160.2c0 2.7-.2 5.4-.5 8.1V472c0 22.1-17.9 40-40 40H456c-1.1 0-2.2 0-3.3-.1c-1.4 .1-2.8 .1-4.2 .1H416 392c-22.1 0-40-17.9-40-40V448 384c0-17.7-14.3-32-32-32H256c-17.7 0-32 14.3-32 32v64 24c0 22.1-17.9 40-40 40H160 128.1c-1.5 0-3-.1-4.5-.2c-1.2 .1-2.4 .2-3.6 .2H104c-22.1 0-40-17.9-40-40V360c0-.9 0-1.9 .1-2.8V287.6H32c-18 0-32-14-32-32.1c0-9 3-17 10-24L266.4 8c7-7 15-8 22-8s15 2 21 7L564.8 231.5c8 7 12 15 11 24z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="flex aspect-square w-full min-w-[50px] items-center justify-center rounded-full border-2 bg-white hover:cursor-pointer lg:min-w-[60px]"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="flex max-h-screen flex-grow flex-col bg-color_2"
      >
        <div
          class="bg-secondary flex w-full flex-row gap-x-4 border-b-4 border-black p-5 grid-in-sidebar-header"
        >
          <div
            class="flex items-center justify-center"
          >
            <div
              class="aspect-square h-[60px] 2xl:h-[80px] overflow-hidden rounded-full bg-white"
            >
              <img
                alt=" picture"
                class="h-full w-full"
                src=""
              />
            </div>
          </div>
          <div
            class="flex flex-col justify-center overflow-hidden"
          >
            <h1
              class="text-outline xl:text-1xl overflow-ellipsis font-play font-bold uppercase text-white md:max-w-[10ch] md:overflow-hidden md:overflow-ellipsis md:text-xl"
            />
            <button
              class="w-max font-play uppercase text-white"
            >
              Edit Profile
            </button>
          </div>
        </div>
        <div
          class="flex flex-col overflow-y-scroll p-4 scrollbar-none md:flex-grow"
        >
          <h2
            class="font-play text-3xl text-white"
          >
            Messages
          </h2>
          <form
            class="mt-2 w-full rounded-xl border-2 border-black bg-color_4 p-3"
          >
            <div
              class="relative w-full"
            >
              <input
                class="rounded-1 w-full border-2 px-1 font-play focus:outline-none"
              />
              <button
                class="absolute right-2 top-1/2 w-fit -translate-y-1/2"
                type="submit"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-magnifying-glass "
                  data-icon="magnifying-glass"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </form>
          <div
            class="mt-5"
          />
        </div>
      </div>
    </aside>
    <div
      class="hidden max-h-screen overflow-x-auto overflow-y-scroll bg-color_2 p-5 scrollbar-none 2xl:flex 2xl:flex-col"
    >
      <h1
        class="text-3xl text-white"
      >
        Users
      </h1>
      <div
        class="mt-5"
      />
    </div>
  </div>
</div>
`;

exports[`App > Loading Screen > renders Loading screen if loading state is true 1`] = `
<div>
  <div
    class="relative min-h-screen w-full bg-color_3"
  >
    <div
      class="absolute left-1/2 top-1/2 -translate-x-1/2 -translate-y-1/2"
    >
      <div
        aria-busy="true"
        aria-label="loading"
        data-testid="audio-loading"
        role="status"
        style="display: flex;"
      >
        <svg
          data-testid="audio-svg"
          fill="white"
          height="80"
          viewBox="0 0 55 80"
          width="80"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            transform="matrix(1 0 0 -1 0 80)"
          >
            <rect
              height="20"
              rx="3"
              width="10"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="4.3s"
                repeatCount="indefinite"
                values="20;45;57;80;64;32;66;45;64;23;66;13;64;56;34;34;2;23;76;79;20"
              />
            </rect>
            <rect
              height="80"
              rx="3"
              width="10"
              x="15"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="2s"
                repeatCount="indefinite"
                values="80;55;33;5;75;23;73;33;12;14;60;80"
              />
            </rect>
            <rect
              height="50"
              rx="3"
              width="10"
              x="30"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="1.4s"
                repeatCount="indefinite"
                values="50;34;78;23;56;23;34;76;80;54;21;50"
              />
            </rect>
            <rect
              height="30"
              rx="3"
              width="10"
              x="45"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="2s"
                repeatCount="indefinite"
                values="30;45;13;80;56;72;45;76;34;23;67;30"
              />
            </rect>
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
